keps/sig-node/4639-oci-volume-source/README.md

```diff
@@ -0,0 +1,940 @@
+- Performance considerations must be taken into account, especially for large images or artifacts.
+
+### Vocabulary: OCI Images, Artifacts, and Objects
+
+**OCI Image ([spec](https://github.com/opencontainers/image-spec/blob/main/spec.md)):**
+   - A container image that conforms to the Open Container Initiative (OCI) Image Specification.
+     It includes a filesystem bundle and metadata required to run a container.
+   - Consists of multiple layers (each layer being a tarball), a manifest (which lists the layers), and a config file
+     (which provides configuration data such as environment variables, entry points, etc.).
+   - **Use Case:** Used primarily for packaging and distributing containerized applications.
+
+**OCI Artifact ([guidance](https://github.com/opencontainers/image-spec/blob/main/artifacts-guidance.md)):**
+   - An artifact describes any content that is stored and distributed using the OCI image format.
+     It includes not just container images but also other types of content like Helm charts, WASM modules, machine learning models, etc.
+   - Artifacts use the same image manifest and layer structure but may contain different types of data
+     within those layers. The artifact manifest can have media types that differ from those in standard container images.
+   - **Use Case:** Allows the distribution of non-container content using the same infrastructure and tools developed for OCI images.
+
+**OCI Object:**
+   - Umbrella term encompassing both OCI images and OCI artifacts. It represents any object that conforms to the OCI specifications for storage and distribution.
+
+### Risks and Mitigations
+
+- **Security Risks:** Allowing direct mounting of OCI images introduces potential attack vectors. Mitigation includes thorough security reviews and
+  limiting access to trusted registries. Limiting to OCI artifacts (non-runnable content) or read-only mode may lessen the security risk.
+- **Compatibility Risks:** Existing webhooks watching for the images used by the pod using some policies will need to be updated to expect the image to be specified as a `VolumeSource`.
+- **Performance Risks:** Large images or artifacts could impact performance. Mitigation includes optimizations in the implementation and providing
+  guidance on best practices for users.
+
+## Design Details
+
+The new `VolumeSource` will be defined in the Kubernetes API, and the implementation will involve updating components (CRI, Kubelet)
+to support this source type. Key design aspects include:
+
+- API changes to introduce the new `VolumeSource` type.
+- Modifications to the Kubelet to handle mounting OCI images and artifacts.
+- Handling image pull secrets and registry authentication.
+- Reuse existing logic from ConfigMaps for:
+  - Determining the file location on the host filesystem.
+  - Handling updates to the OCI image or artifact, similar to how ConfigMaps update mounted files.
```

Comment:
This one is a really interesting topic.  IMO the fact that kube live-updates configmaps by default kind of goes against the whole immutable-infrastructure model.

I am curious what others think but I would argue that live update support should be an opt-in, [not an opt-out](https://kubernetes.io/docs/concepts/configuration/configmap/#configmap-immutable).

Reply:
> ConfigMaps you can watch, whereas there's no equivalent for OCI. I can imagine it but I don't think it exists.

In OpenShift there is the [concept of an image stream](https://docs.openshift.com/container-platform/4.15/rest_api/image_apis/imagestream-image-openshift-io-v1.html), but yeah not in upstream Kube AFAIK.

Reply:
BTW I don't know if we should push for this feature in alpha it seems complex

Reply:
Agree that this should be left out and can be tackled later if we find use compelling uses cases or patterns that can't be solved without this.

Reply:
What is the expectation of the reverse, rather than external changes pushed to the container, how do changes from the container get handled. Is the volume read-only, are writes ephemeral and discarded when the pod is removed, or are changes permitted and propagated back up somewhere?

Reply:
> Is the volume read-only, 

Please yes.  Let's not repeat the sins of docker here in defaulting to a writable, ephemeral mount.  That's just an attractive trap.